pkg/kv/kvserver/store_send.go

```diff
@@ -157,104 +158,166 @@ func (s *Store) Send(
+			ri, err := t.MismatchedRange()
+			if err != nil {
+				return nil, roachpb.NewError(err)
+			}
+			skipRID := ri.Desc.RangeID // We already have info on one range, so don't add it again below.
+			startKey := ri.Desc.StartKey
+			if rSpan.Key.Less(startKey) {
+				startKey = rSpan.Key
+			}
+			endKey := ri.Desc.EndKey
+			if endKey.Less(rSpan.EndKey) {
+				endKey = rSpan.EndKey
+			}
+			var ris []roachpb.RangeInfo
+			if err := s.visitReplicasByKey(ctx, startKey, endKey, AscendingKeyOrder, func(ctx context.Context, repl *Replica) error {
+				// Note that we return the lease even if it's expired. The kvclient can use it as it sees fit.
+				ri := repl.GetRangeInfo(ctx)
+				if ri.Desc.RangeID == skipRID {
+					return nil
+				}
+				ris = append(ris, ri)
 				return nil
+			}); err != nil {
+				// Errors here should not be possible, but if there is one, it is ignored
+				// as attaching RangeInfo is optional.
+				log.Warningf(ctx, "unexpected error visiting replicas: %s", err)
+				ris = nil // just to be safe
 			}
-			ris = append(ris, ri)
-			return nil
-		}); err != nil {
-			// Errors here should not be possible, but if there is one, it is ignored
-			// as attaching RangeInfo is optional.
-			log.Warningf(ctx, "unexpected error visiting replicas: %s", err)
-			ris = nil // just to be safe
-		}
-		for _, ri := range ris {
-			t.AppendRangeInfo(ctx, ri)
+			isRetriableRangeKeyMismatch := false
+			candidateRanges = append(candidateRanges, t.Ranges...)
+			for _, ri := range ris {
+				// Update the candidate ranges, that will be combined with the replica batch response RangeInfos
+				// if returned from the replica. Note here that newer ranges are always appended, so that the
+				// oldest RangeInfo is processed by the Client's RangeCache first, which is then invalidated on
+				// conflict by newer data (LILO).
+				ri.ValidateRangeInfoLease(ctx)
+				candidateRanges = append(candidateRanges, ri)
+
+				// Check if the original batch request rSpan exists entirely within any known
+				// ranges stored on this replica, if so update repl and set retriable to true.
+				if ri.Desc.RSpan().ContainsKeyRange(rSpan.Key, rSpan.EndKey) {
+					// Retry this request: Update BatchRequest to reflect re-routing the request
+					// to ri. ClientRangeInfo is also updated here to avoid the replica providing
+					// duplicate BatchResponse.RangeInfos upon a successful retry for the client
+					// to invalidate their RangeCache with.
+					isRetriableRangeKeyMismatch = true
+
+					// Refresh the request and populate the updated request information.
+					baRetry = ba
+					baRetry.RangeID = ri.Desc.RangeID
+					baRetry.ClientRangeInfo.ClosedTimestampPolicy = ri.ClosedTimestampPolicy
+					baRetry.ClientRangeInfo.DescriptorGeneration = ri.Desc.Generation
+					baRetry.ClientRangeInfo.LeaseSequence = ri.Lease.Sequence
+				}
+			}
+
+			// if we are able to retry, iterate retry loop.
+			if isRetriableRangeKeyMismatch {
```

Comment:
At the cost of adding another loop over `ris`, we could do the following which I think would be easier to read + avoid the need for `isRetriableRangeKeyMismatch`.

```go
for _, ri := range ris {
  candidateRanges = append(candidateRanges, ri)
}

for _, ri := range ris {
  if ri.Desc.RSpan().ContainsKeyRange(rSpan.Key, rSpan.EndKey) {
    baRetry = ba // ... and other stuff
    candidateRanges = append(candidateRanges, t.Ranges...) // collect for next loop iteration, instead of where we're doing it now
    continue
  }
}
```

In fact, the first loop is actually a one liner.
```go
candidateRanges = append(candidateRanges, ris...)
```

[nit] I'd also suggest s/candidateRanges/rangeInfos for two reasons:
- They're not ranges, which are a distinct separate thing
- They're not candidates, we're going to unconditionally return everything in the slice, whether through the batch response or through pErr.

Reply:
I like the idea of splitting. The only issue I have is I don't think it is possible to nested continue here. Instead, it will just continue the inner loop. Otherwise, I'll add.